testing/scenario_app/ios/Scenarios/ScenariosUITests/iPadGestureTests.m

```diff
@@ -62,23 +69,123 @@ - (void)testPointerButtons {
+      @"PointerChange.down event did not occur for a normal tap");
+  XCTAssertTrue(
+      [app.textFields[@"2,PointerChange.up,device=0,buttons=0"] waitForExistenceWithTimeout:1],
+      @"PointerChange.up event did not occur for a normal tap");
   SEL rightClick = @selector(rightClick);
   XCTAssertTrue([flutterView respondsToSelector:rightClick],
                 @"If supportsPointerInteraction is true, this should be true too.");
   [flutterView performSelector:rightClick];
-  // Since each touch is its own device, we can't distinguish the other add event(s)
+  // On simulated right click, a hover also occurs, so the hover pointer is added
+  XCTAssertTrue(
+      [app.textFields[@"3,PointerChange.add,device=1,buttons=0"] waitForExistenceWithTimeout:1],
+      @"PointerChange.add event did not occur for a right-click's hover pointer");
+  [NSThread sleepForTimeInterval:5.0];  // The hover pointer is removed after ~3.5 seconds, this
+                                        // ensures all events have been received
+  // The hover events are interspersed with the right-click events in a varying order
+  // Ensure the individual orderings are respected without hardcoding the absolute sequence
+  NSMutableDictionary* messages = [NSMutableDictionary dictionary];
+  for (XCUIElement* element in [app.textFields allElementsBoundByIndex]) {
+    NSString* rawMessage = element.value;
+    NSUInteger commaIndex = [rawMessage rangeOfString:@","].location;
+    NSInteger messageSequenceNo =
+        [[rawMessage substringWithRange:NSMakeRange(0, commaIndex)] integerValue];
+    NSString* message = [rawMessage
+        substringWithRange:NSMakeRange(commaIndex + 1, [rawMessage length] - (commaIndex + 1))];
+    NSMutableArray* messageSequenceNoList = [messages objectForKey:message];
+    if (messageSequenceNoList == nil) {
+      messageSequenceNoList = [[NSMutableArray alloc] init];
+      [messages setObject:messageSequenceNoList forKey:message];
+    }
+    [messageSequenceNoList addObject:[NSNumber numberWithInteger:messageSequenceNo]];
+  }
+  // The number of hover events is not consistent
+  NSMutableArray* hoverSeqNos = [messages objectForKey:@"PointerChange.hover,device=1,buttons=0"];
+  int hoverRemovedSeqNo =
+      assertOneMessageAndGetSequenceNo(messages, @"PointerChange.remove,device=1,buttons=0");
   // Right click should have buttons = 2
-  XCTAssertTrue([app.textFields[@"PointerChange.down:2"] waitForExistenceWithTimeout:1],
-                @"PointerChange.down event did not occur for a right-click");
-  XCTAssertTrue([app.textFields[@"PointerChange.up:2"] waitForExistenceWithTimeout:1],
-                @"PointerChange.up event did not occur for a right-click");
+  int rightClickAddedSeqNo, rightClickDownSeqNo, rightClickUpSeqNo;
+  if ([messages objectForKey:@"PointerChange.add,device=2,buttons=0"] == nil) {
+    // Sometimes the tap pointer has the same device as the right-click (the UITouch is reused)
```

Comment:
This test generally may be easier to read as a series of expectations, with ordering either enforced or not depending on what you're checking.

Reply:
I observed a lot of different permutations of event order, even though this is somewhat convoluted I don't think the alternative is better.